File: Java/LinkedListPractice/src/LinkedList/LinkedListTest.java
package LinkedList;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.util.Optional;
import java.util.function.BooleanSupplier;

import static org.junit.jupiter.api.Assertions.*;

class LinkedListTest {

    private final PrintStream standardOut = System.out;
    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
    LinkedList testLinkedList;

    @BeforeEach
    void setUp() {
        System.setOut(new PrintStream(outputStreamCaptor));
        this.testLinkedList = new LinkedList(10);
        this.testLinkedList.add(100);
    }

    @AfterEach
    public void tearDown() {
        System.setOut(standardOut);
    }

    @Test
    void add() {
        this.testLinkedList.add(200);
        assertEquals(this.testLinkedList.tail.getData(), 200);
    }

    @Test
    void contains() {
        assertTrue(this.testLinkedList.contains(100));
    }

    @Test
    void get() {
        Optional<Integer> optional = this.testLinkedList.get(0);
        assertTrue(optional.isPresent());
        assertEquals(100, optional.get());
    }

    @Test
    void insertAt() {
        this.testLinkedList.insertAt(0, 50);
        assertEquals(50, this.testLinkedList.head.getData());
    }

    @Test
    void printAll() {
        this.testLinkedList.add(400);
        this.testLinkedList.printAll();
        String output = outputStreamCaptor.toString().trim();
        assertTrue(output.contains("100"));
        assertTrue(output.contains("400"));
    }

    @Test
    void removeFrom() {
        this.testLinkedList.removeFrom(0);
        assertFalse(this.testLinkedList.contains(100));
    }
}

File: Kotlin/LinkedListPractice/test/linkedList/LinkedListTest.kt
package linkedList

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import java.io.ByteArrayOutputStream
import java.io.PrintStream

class LinkedListTest {

    private val linkedList = LinkedList(head = null, tail = null)
    private val outputStreamCaptor = ByteArrayOutputStream()

    @BeforeEach
    fun setUp() {
        System.setOut(PrintStream(outputStreamCaptor));
        linkedList.add(1)
        linkedList.add(2)
    }

    @org.junit.jupiter.api.Test
    fun add() {
        linkedList.head?.let { assertEquals(it.data, 1) }
        linkedList.tail?.let { assertEquals(it.data, 2) }
    }

    @org.junit.jupiter.api.Test
    fun contains() {
        assertTrue(linkedList.contains(1))
    }

    @org.junit.jupiter.api.Test
    fun get() {
        linkedList.get(1)?.let { assertEquals(it.data, 1) }
    }

    @org.junit.jupiter.api.Test
    fun getAtIndex() {
        linkedList.getAtIndex(1)?.let { assertEquals(it.data, 2) }
        linkedList.getAtIndex(0)?.let { assertEquals(it.data, 1) }
    }

    @org.junit.jupiter.api.Test
    fun insertAt() {
        linkedList.insertAt(1, 3)
        assertTrue(linkedList.contains(3))
        linkedList.get(3)?.let { assertEquals(it.data, 3) }
        linkedList.tail?.let { assertEquals(it.data, 2) }
    }

    @org.junit.jupiter.api.Test
    fun length() {
        assertEquals(linkedList.length(), 2)
    }

    @org.junit.jupiter.api.Test
    fun printAll() {
        this.linkedList.printAll()
        val output = outputStreamCaptor.toString().trim { it <= ' ' }
        assertTrue(output.contains("1"))
        assertTrue(output.contains("2"))
    }

    @org.junit.jupiter.api.Test
    fun remove() {
        linkedList.remove(1)
        assertFalse(linkedList.contains(2))
        linkedList.tail?.let { assertEquals(it.data, 1) }
    }
}